bootstrap/package_manager.py

```diff
@@ -0,0 +1,325 @@
+
+    def _get_installed_packages(self) -> Set[str]:
+        raise Exception("NotImplementedException")
+
+    def _filter_out_installed_packages(self, dependencies: Dict[str, Set[str]]):
+        installed_packages = self._get_installed_packages()
+        filtered_packages = {k: (v - installed_packages) for k, v in dependencies.items()}
+        for installed_package in installed_packages:
+            filtered_packages.pop(installed_package, None)
+        return filtered_packages
+
+    def run_cmd(self, cmd: str) -> bool:
+        result = subprocess.run(f"{cmd}", shell=True, text=True)
+        return result.returncode == 0
+
+
+class BrewPackageManager(PackageManager):
+    def __init__(self, no_confirm):
+        PackageManager.__init__(self, no_confirm)
+
+    def install(self, dependencies: Dict[str, Set[str]]) -> bool:
+        return self._install(dependencies=dependencies,
+                             install_cmd="brew install",
+                             replace_dict={"patch": set(),
+                                           "jni": {"maven"}})
+
+    def install_compiler(self) -> str:
+        self.install({"compiler": {"llvm"}})
+        return ""
+
+    def _get_installed_packages(self) -> Set[str]:
+        result = subprocess.run(['brew', 'list'], text=True, capture_output=True, check=True)
+        lines = result.stdout.splitlines()
+        lines = [line.split('@', 1)[0] for line in lines]
+        return set(lines)
+
+
+class AptPackageManager(PackageManager):
+    def __init__(self, no_confirm):
+        PackageManager.__init__(self, no_confirm)
+
+    def install(self, dependencies: Dict[str, Set[str]]) -> bool:
+        return self._install(dependencies=dependencies,
+                             install_cmd="sudo apt install -y",
+                             replace_dict={"libarchive": {"liblzma-dev"},
+                                           "lua": {"liblua5.1-0-dev"},
+                                           "python": {"libpython3-dev"},
+                                           "libusb": {"libusb-1.0-0-dev", "libusb-dev"},
+                                           "libpng": {"libpng-dev"},
+                                           "libpcap": {"libpcap-dev"},
+                                           "jni": {"openjdk-8-jdk", "openjdk-8-source", "maven"},
+                                           "gpsd": {"libgps-dev"}})
+
+    def _get_installed_packages(self) -> Set[str]:
+        result = subprocess.run(['dpkg', '--get-selections'], text=True, capture_output=True, check=True)
+        lines = [line.split('\t')[0] for line in result.stdout.splitlines()]
+        lines = [line.rsplit(':', 1)[0] for line in lines]
+        return set(lines)
+
+    def install_compiler(self) -> str:
+        if distro.id() == "ubuntu" and int(distro.major_version()) < 22:
+            self.install({"compiler_prereq": {"software-properties-common"}})
```

Comment:
Most 3rd party repos also need: `apt-transport-https ca-certificates software-properties-common`

It may be more noticable with a minimal server install or docker image, but I didn't test yet.

Reply:
👍 https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/6b7d1af1293bf49e121d32f90e6cd9617c78a68e#diff-0ebcdc25be3be8a77e8b404eb6962891695836a8ca6bfa65607a2622d114f35dR125